mesonbuild/environment.py

```diff
@@ -371,17 +372,24 @@ def __init__(self, source_dir, build_dir, options):
         # Similar to coredata.compilers and build.compilers, but lower level in
         # that there is no meta data, only names/paths.
         self.binaries = PerMachineDefaultable()
+
         # Just uses hard-coded defaults and environment variables. Might be
         # overwritten by a native file.
         self.binaries.build = BinaryTable({})
 
         # Misc other properties about each machine.
         self.properties = PerMachine(Properties(), Properties(), Properties())
 
+        # Store paths for native and cross build files. There is no target
+        # machine information here because nothing is installed for the target
+        # architecture, just the build and host architectures
+        self.paths = PerMachine(Directories(), Directories(), None)
```

Comment:
Well by that logic this shouldn't be a `PerMachine` because nothing is ever installed for the build platform by definition. (It is the host platform, and when build == host, it is because of the host platform, not because of the build platform aliasing it.)

All things target-platform related are dead code in Meson today, btw, but if/when Meson gets used for some compiler, it may be nice to specify per-target paths separately. (Consider multilib GCC or LLVM and their `/lib/compiler-config` directories.)

Reply:
If you wish to keep `PerMachine` for future target support, use `PerMachineDefaultable` so that you can default `paths.host` when no cross file is used, to reflect that the host and build configurations are the same.

Reply:
Sure, but the nice thing about storing it in a PerMachine is that we don't have to special case the cross/native reading machinery. We just load the cross and the native files and then use the right one.

Reply:
Also, that's not necessarily true. LLVM can build llvm-config for the build system, even when libLLVM.so is compiled for the host system. So it is possible that you could need both.

Reply:
Hehe, `*-config` and friends are indeed an exception to my rule, and one that's caused us at my distro a ton of pain! https://github.com/NixOS/nixpkgs/issues/51176

Reply:
Sure -config has caused **me** no end of pain. Unfortunately no one asks me :)